pylint/testutils/functional/find_functional_tests.py

```diff
@@ -50,27 +51,38 @@ def get_functional_test_files_from_directory(
 
 
 def _check_functional_tests_structure(directory: Path) -> None:
-    """Check if test directories follow correct file/folder structure."""
-    # Ignore underscored directories
+    """Check if test directories follow correct file/folder structure.
+
+    Ignore underscored directories or files.
+    """
     if Path(directory).stem.startswith("_"):
         return
 
     files: set[Path] = set()
     dirs: set[Path] = set()
 
+    def walk(path: Path) -> Iterator[Path]:
+        for _file_or_dir in path.iterdir():
+            if _file_or_dir.is_dir():
+                _files = list(_file_or_dir.iterdir())
+                assert len(_files) <= REASONABLY_DISPLAYABLE_VERTICALLY, (
+                    f"{_file_or_dir} contains too many functional tests files "
+                    + f"({len(_files)} > {REASONABLY_DISPLAYABLE_VERTICALLY})."
+                )
+                yield _file_or_dir
+                yield from walk(_file_or_dir)
+            else:
+                yield _file_or_dir.resolve()
+
     # Collect all sub-directories and files in directory
-    for file_or_dir in directory.iterdir():
-        if file_or_dir.is_file():
-            if file_or_dir.suffix == ".py" and not file_or_dir.stem.startswith("_"):
-                files.add(file_or_dir)
-        elif file_or_dir.is_dir():
+    for file_or_dir in walk(directory):
+        if file_or_dir.stem.startswith("_"):
```

Comment:
We should probably do the check for directory length after excluding ignored files

Reply:
[5bee0d6](https://github.com/PyCQA/pylint/pull/8484/commits/5bee0d67db163a463d861181820a3759fb32e070)